### PR TITLE
MTDSA-17558: Enabling ability for the passage of intent header through the api

### PR DESCRIPTION
**Pull request by harry-ritchie**

### CODE REVIEW

File: app/v1/controllers/AddCustomEmploymentController.scala [hunk 1]
```diff
@@ -57,7 +57,7 @@ class AddCustomEmploymentController @Inject() (val authService: EnrolmentsAuthSe
         nino = nino,
         taxYear = taxYear,
         body = AnyContentAsJson(request.body),
-        temporalValidationEnabled = FeatureSwitches()(appConfig).isTemporalValidationEnabled)
+        temporalValidationEnabled = FeatureSwitches(appConfig.featureSwitches).isTemporalValidationEnabled(request))
```
Comment (jeremystone) [inline app/v1/controllers/AddCustomEmploymentController.scala:60]: As `request` is implicit:
```suggestion
        temporalValidationEnabled = FeatureSwitches(appConfig.featureSwitches).isTemporalValidationEnabled))
```

File: test/v1/controllers/AddCustomEmploymentControllerSpec.scala [hunk 2]
```diff
@@ -45,7 +45,8 @@ class AddCustomEmploymentControllerSpec
     with MockAddCustomEmploymentService
     with MockAuditService
     with MockAddCustomEmploymentRequestParser
-    with MockHateoasFactory {
+    with MockHateoasFactory
+    with MockFeatureSwitches {
```
Comment (jeremystone) [inline test/v1/controllers/AddCustomEmploymentControllerSpec.scala:49]: Is `MockFeatureSwitches` required here? Doesn't look like it is being used.

File: test/api/connectors/ConnectorSpec.scala [hunk 3]
```diff
@@ -187,7 +186,7 @@ trait ConnectorSpec extends UnitSpec with Status with MimeTypes with HeaderNames
     MockedAppConfig.ifsEnvironment returns "ifs-environment"
     MockedAppConfig.ifsEnvironmentHeaders returns Some(allowedIfsHeaders)
 
-    MockedAppConfig.featureSwitches returns Configuration("tys-api.enabled" -> false)
+    MockFeatureSwitches.isTaxYearSpecificApiEnabled.returns(true).anyNumberOfTimes()
```
Comment (jeremystone) [inline test/api/connectors/ConnectorSpec.scala:189]: Was false before; should be false now:
```suggestion
    MockFeatureSwitches.isTaxYearSpecificApiEnabled.returns(false).anyNumberOfTimes()
```

File: test/api/connectors/ConnectorSpec.scala [hunk 4]
```diff
@@ -199,7 +198,7 @@ trait ConnectorSpec extends UnitSpec with Status with MimeTypes with HeaderNames
     MockedAppConfig.release6Environment returns "release6-environment"
     MockedAppConfig.release6EnvironmentHeaders returns Some(allowedIfsHeaders)
 
-    MockedAppConfig.featureSwitches returns Configuration("tys-api.enabled" -> false)
+    MockFeatureSwitches.isTaxYearSpecificApiEnabled.returns(true).anyNumberOfTimes()
```
Comment (jeremystone) [inline test/api/connectors/ConnectorSpec.scala:201]: Was false before; should be false now:
```suggestion
    MockFeatureSwitches.isTaxYearSpecificApiEnabled.returns(false).anyNumberOfTimes()
```

File: test/api/connectors/ConnectorSpec.scala [hunk 5]
```diff
@@ -211,7 +210,7 @@ trait ConnectorSpec extends UnitSpec with Status with MimeTypes with HeaderNames
     MockedAppConfig.api1661Environment returns "api1661-environment"
     MockedAppConfig.api1661EnvironmentHeaders returns Some(allowedIfsHeaders)
 
-    MockedAppConfig.featureSwitches returns Configuration("tys-api.enabled" -> false)
+    MockFeatureSwitches.isTaxYearSpecificApiEnabled.returns(true).anyNumberOfTimes()
```
Comment (jeremystone) [inline test/api/connectors/ConnectorSpec.scala:213]: Was false before; should be false now:
```suggestion
    MockFeatureSwitches.isTaxYearSpecificApiEnabled.returns(false).anyNumberOfTimes()
```

File: test/api/connectors/ConnectorSpec.scala [hunk 6]
```diff
@@ -174,8 +174,7 @@ trait ConnectorSpec extends UnitSpec with Status with MimeTypes with HeaderNames
     MockedAppConfig.desEnvironment returns "des-environment"
     MockedAppConfig.desEnvironmentHeaders returns Some(allowedDesHeaders)
 
-    MockedAppConfig.featureSwitches returns Configuration("tys-api.enabled" -> false)
-
+    MockFeatureSwitches.isTaxYearSpecificApiEnabled.returns(true).anyNumberOfTimes()
```
Comment (jeremystone) [inline test/api/connectors/ConnectorSpec.scala:177]: Was false before; should be false now:
```suggestion
    MockFeatureSwitches.isTaxYearSpecificApiEnabled.returns(false).anyNumberOfTimes()
```

File: app/v1/connectors/DeleteUkDividendsIncomeAnnualSummaryConnector.scala [hunk 7]
```diff
@@ -37,14 +39,18 @@ class DeleteUkDividendsIncomeAnnualSummaryConnector @Inject() (val http: HttpCli
 
     import request._
 
+    val intent = if (featureSwitches.isPassDeleteIntentEnabled) Some("DELETE") else None
+
     if (taxYear.useTaxYearSpecificApi) {
       delete(
-        uri = TaxYearSpecificIfsUri[Unit](s"income-tax/${taxYear.asTysDownstream}/${nino.nino}/income-source/dividends/annual")
+        uri = TaxYearSpecificIfsUri[Unit](s"income-tax/${taxYear.asTysDownstream}/${nino.nino}/income-source/dividends/annual"),
+        intent
```
Comment (jeremystone) [inline app/v1/connectors/DeleteUkDividendsIncomeAnnualSummaryConnector.scala:47]: Not sure we need to pass a DELETE intent when it is already a delete call. (Might simplify changes need to be made to the stub.)

File: test/v1/connectors/DeleteUkDividendsIncomeAnnualSummaryConnectorSpec.scala [hunk 8]
```diff
@@ -37,17 +39,51 @@ class DeleteUkDividendsIncomeAnnualSummaryConnectorSpec extends ConnectorSpec {
...
+        ).returns(Future.successful(outcome))
+
+        MockFeatureSwitches.isPassDeleteIntentEnabled.returns(false)
+
         await(connector.delete(request)) shouldBe outcome
       }
     }
     "return the expected response for a TYS request" when {
-      "a valid request is made" in new TysIfsTest with Test {
+      "isPassDeleteIntentEnabled feature switch is on" in new TysIfsTest with Test {
```
Comment (jeremystone) [inline test/v1/connectors/DeleteUkDividendsIncomeAnnualSummaryConnectorSpec.scala:63]: See comment in connector -- might not need to pass intent at all for something that is already a delete.

File: test/api/connectors/ConnectorSpec.scala (hunk 6 above)
Comment (harry-ritchie) [reply test/api/connectors/ConnectorSpec.scala:177]: Good catch.

File: app/v1/connectors/DeleteUkDividendsIncomeAnnualSummaryConnector.scala (hunk 7 above)
Comment (harry-ritchie) [reply app/v1/connectors/DeleteUkDividendsIncomeAnnualSummaryConnector.scala:47]: Yes you are right!

**Review by jeremystone — DISMISSED** [review]

**Review by jeremystone — APPROVED** [review]

**Pull request merged**